indico/modules/users/util.py

```diff
@@ -419,6 +421,57 @@ def merge_users(source, target, force=False):
     logger.info('Successfully merged %s into %s', source, target)
 
 
+def anonymize_user(user):
+    from indico.modules.events.models.principals import EventPrincipal
+    from indico.modules.events.models.settings import EventSettingPrincipal
+    from indico.modules.events.sessions.models.principals import SessionPrincipal
+    from indico.modules.events.tracks.models.principals import TrackPrincipal
+    from indico.modules.rb.models.blocking_principals import BlockingPrincipal
+    from indico.modules.rb.models.principals import RoomPrincipal
+
+    user.first_name = '<anonymous>'
+    user.last_name = f'<anonymous-{user.id}>'
+    user.title = UserTitle.none
+    user.affiliation = ''
+    user.affiliation_id = None
+    user.phone = ''
+    user.address = ''
+    user.picture = None
+    user.picture_metadata = None
+    user.picture_source = ProfilePictureSource.standard
+    user.favorite_users = set()
+    user.favorite_events = set()
+    user.favorite_categories = set()
+    user.identities = set()
+
+    UserEmail.query.filter(UserEmail.user == user).delete()
+    email = UserEmail(user_id=user.id, email=f'indico-{user.id}@indico.invalid')
+    db.session.add(email)
+
+    editables = user.editor_for_editables
+    for editable in editables:
+        editable.editor = None
+
+    reservations = user.reservations_booked_for
+    for reservation in reservations:
+        reservation.booked_for_name = user.full_name
+
+    categories = user.category_roles.copy()
+    for category in categories:
+        category.members.remove(user)
+
+    principals = [
+        EventPrincipal, CategoryPrincipal, AttachmentPrincipal, AttachmentFolderPrincipal,
+        BlockingPrincipal, TrackPrincipal, EventSettingPrincipal, SettingPrincipal, RoomPrincipal,
+        SessionPrincipal
+    ]
```

Comment:
This should not be hardcoded. You can use this to get a list of all principal classes:

```py
    classes = [sc for sc in [*PrincipalMixin.__subclasses__(), *PrincipalPermissionsMixin.__subclasses__()]
               if hasattr(sc, 'query')]
```